### PR TITLE
invoices: do not log when invoice is already deleted

### PR DESCRIPTION
**Pull request by bhandras**
fixes https://github.com/lightningnetwork/lnd/issues/5911

### CODE REVIEW

File: invoices/invoice_expiry_watcher.go [hunk 1]
```diff
@@ -353,6 +353,10 @@ func (ew *InvoiceExpiryWatcher) expireInvoice(hash lntypes.Hash, force bool) {
 
 	case channeldb.ErrInvoiceAlreadySettled:
 
+	case channeldb.ErrInvoiceNotFound:
+		// It's possible that the user already canceled the invoice
+		// which then will be deleted by the time we'll try to cancel.
```
Comment (guggero) [inline invoices/invoice_expiry_watcher.go:358]: nit: mention the auto garbage collection here as well since that's the cause of this?

**Review by guggero — APPROVED** [review]
Thanks for the quick fix! LGTM :tada:

File: invoices/invoice_expiry_watcher.go (hunk 1 above)
Comment (bhandras) [reply invoices/invoice_expiry_watcher.go:358]: done

**Review by positiveblue — APPROVED** [review]
LGTM but this could potentially hide some bugs.

If we are watching one invoice and when we want to mark it as expired it does not exist it can be because two things:

- Something went wrong
- We should not be trying to mark that invoice as expired

I think it would make more sense to stop watching invoices when we cancel them. So if the watcher tries to expire one invoice and it does not exist, we know it is because something went wrong.

**Comment by bhandras** [comment]
> If we are watching an invoice and when we want to mark it as expired it does not exist it can be because two things:

> Something went wrong
> We should not be trying to mark that invoice as expired
> I think it would make more sense to stop watching invoices when we cancel them. So if the watcher tries to expire one > invoice and it does not exist, we know it is because something went wrong.

Yes generally agree with your remarks, however please note that the expiry watcher doesn't hold any invoice state apart from the invoice hash and the expiry. The idea is instead to do all invoice manipulation inside the registry and only use the watcher as a "ticker" when the registry should try and cancel expired invoices.

So the flow roughly is:
1) registry adds an invoice (hash, expiry) to the watcher
2) watcher notifies registry about expiration
3) registry attempts to cancel and delete (if requested)

**Comment by positiveblue** [comment]
@bhandras I totally get it. Here it is also easier to just stop logging because there is no way to "delete" an invoice from the watcher (it is a priority queue with only the top/pop interface methods and not a remove).

If we wanted to stop notifying the registry about "false" expiration we should keep a complementary map with the "real" tracked hashes and have a `func (ew *InvoiceExpiryWatcher) DeletInvoices(invoices ...invoiceExpiry)`. Whenever we are taking an item from the queue if it is not in the map it means that we stopped tracking it, so we can just `pass`.

I was just dumping my thoughts there, that's why I approved the PR too :)

**Pull request merged**